### PR TITLE
Only create aoMetRoughTex when metalnesss or roughness texture provided

### PR DESCRIPTION
**Pull request by mtostenson**
This change prevents the creation of the combined metalness and roughness texture for PBR FBX materials if neither a metalness or roughness map has been provided.

The motivation for this change was to prevent overriding the provided constant values for metalness and roughness. In our situation, we are not using metalness or roughness maps and provide constants for both values, but they were being lost because the combined map was being created anyway. In the file `MaterialData.cpp` the `to_json()` method forces the roughness and metalness values to 1 if the metRoughTexture exists, so it should follow that the texture should only be created if we provide either a metalness or roughness map from the beginning.

### CODE REVIEW

**Comment by zellski** [comment]
Oh good lord, it looks like we always did the wrong thing here. Many thanks. I'll run it through a couple of test suites tonight & then happily merge.

File: src/gltf/Raw2Gltf.cpp [hunk 1]
```diff
@@ -261,26 +261,31 @@ ModelData* Raw2Gltf(
...
-                         metallic,
-                         1}};
-              },
-              false);
-
+          
+          // If either a metallic or roughness map has been provided, generate the aoMetRoughTex.
+          bool hasMetallicMap = material.textures[RAW_TEXTURE_USAGE_METALLIC] >= 0;
+          bool hasRoughnessMap = material.textures[RAW_TEXTURE_USAGE_ROUGHNESS] >= 0;
+          if (hasMetallicMap || hasRoughnessMap) {
```
Comment (zellski) [inline src/gltf/Raw2Gltf.cpp]: This should have a hasOcclusionMap check as well, right? As unlikely as it is that a model has an AO map without either metallic or roughness maps, it seems like the right thing to do.

Comment (mtostenson) [reply src/gltf/Raw2Gltf.cpp]: I'm not a 3d modeler myself, but I know that we this is a common situation for us. We have a few thousand models that use AO maps but only constants for metalness and roughness. In the tests I've done there haven't been any issues supplying only an AO map and not generating metal/roughness map.

Comment (zellski) [reply src/gltf/Raw2Gltf.cpp]: Hmm. Confusing. It works with such a combination, even with this change? I don’t set how occlusionTexture would get set. But in any case I realise now, if there’s an AO map but constant roughness & metal, of course we shouldn’t run the texture rebuilder anyway, but rather just copy the source texture. I’ll merge this in a few hours, apologies for the delay!

**Comment by zellski** [comment]
I merged this –– thanks! After closer perusal of the section in general, I also took it a step further in https://github.com/facebookincubator/FBX2glTF/commit/d2f7d0e270156509f18390244fa17fae65d92712. Would you be able to run this against a couple of models with occlusion maps –– ideally with and without metallic/roughness maps? I realise I don't have any in my test suite (must fix).

File: src/gltf/Raw2Gltf.cpp (hunk 1 above)
Comment (mtostenson) [reply src/gltf/Raw2Gltf.cpp]: Yeah I think the final gltf was actually just using the occlusion map separately anyway, even when it generated the metalness roughness map. I wondered if that wasn't intentional.

**Comment by mtostenson** [comment]
> I merged this –– thanks! After closer perusal of the section in general, I also took it a step further in [d2f7d0e](https://github.com/facebookincubator/FBX2glTF/commit/d2f7d0e270156509f18390244fa17fae65d92712). Would you be able to run this against a couple of models with occlusion maps –– ideally with and without metallic/roughness maps? I realise I don't have any in my test suite (must fix).

I tested this on one of our models that provide a constant for metalness and roughness, and it looks like they got overridden. Here's the material object from the glTF:

```
        {
            "name": "mat_metal",
            "alphaMode": "OPAQUE",
            "extras": {
                "fromFBX": {
                    "shadingModel": "Metallic/Roughness",
                    "isTruePBR": true
                }
            },
            "occlusionTexture": {
                "index": 1,
                "texCoord": 0
            },
            "pbrMetallicRoughness": {
                "baseColorTexture": {
                    "index": 2,
                    "texCoord": 0
                },
                "baseColorFactor": [
                    0.5,
                    0.5,
                    0.5,
                    1.0
                ],
                "metallicRoughnessTexture": {
                    "index": 1,
                    "texCoord": 0
                },
                "roughnessFactor": 1.0,
                "metallicFactor": 1.0
            }
```

**Comment by zellski** [comment]
Ah, crap, yes. More logic errors on my part. Will take another pass in a while.

(If you hadn't guessed, our regression suite is still very poor.)

I don't suppose, for sake of expedience, that you have a FBX file available that you're at liberty to send me?

**Comment by mtostenson** [comment]
> Ah, crap, yes. More logic errors on my part. Will take another pass in a while.
> 
> (If you hadn't guessed, our regression suite is still very poor.)
> 
> I don't suppose, for sake of expedience, that you have a FBX file available that you're at liberty to send me?

@LaurenPozzi do you have something?

**Comment by LaurenPozzi** [comment]
Attached. This includes AO maps and has values plugged in for color, roughness, and metalness. 
[PBR toilet fbx.zip](https://github.com/facebookincubator/FBX2glTF/files/3030770/PBR.toilet.fbx.zip)

**Comment by zellski** [comment]
Fantastic! Thank you. Will sic code at it shortly.

**Comment by mtostenson** [comment]
@zellski Any chance you could kick off a windows build so my artists can get the latest?

**Comment by zellski** [comment]
@mtostenson apologies for the delay –– kicked off builds here: https://dev.azure.com/parwinzell/FBX2glTF/_build/results?buildId=45

**Pull request merged**